zuul-core/src/main/java/com/netflix/zuul/netty/server/push/PushAuthHandler.java

```diff
@@ -0,0 +1,117 @@
+ */
+@ChannelHandler.Sharable
+
+public abstract class PushAuthHandler extends SimpleChannelInboundHandler<FullHttpRequest> {
+
+    private final String pushConnectionPath;
+    private final String originDomain;
+
+    public static final String NAME = "push_auth_handler";
+    private static Logger logger = LoggerFactory.getLogger(PushAuthHandler.class);
+
+
+    public PushAuthHandler(String pushConnectionPath, String originDomain) {
+        this.pushConnectionPath = pushConnectionPath;
+        this.originDomain = originDomain;
+    }
+
+    public final void sendHttpResponse(HttpRequest req, ChannelHandlerContext ctx, HttpResponseStatus status) {
+        FullHttpResponse resp = new DefaultFullHttpResponse(HTTP_1_1, status);
+        resp.headers().add("Content-Length", "0");
+        final boolean closeConn = ((status != OK) || (! HttpUtil.isKeepAlive(req)));
+        if (closeConn)  {
+            resp.headers().add(HttpHeaderNames.CONNECTION, "Close");
+        }
+        final ChannelFuture cf = ctx.channel().writeAndFlush(resp);
+        if (closeConn) {
+            cf.addListener(ChannelFutureListener.CLOSE);
+        }
+    }
+
+    @Override
+    protected final void channelRead0(ChannelHandlerContext ctx, FullHttpRequest req) throws Exception {
+        if (req.method() != HttpMethod.GET) {
+            sendHttpResponse(req, ctx, METHOD_NOT_ALLOWED);
+            return;
+        }
+
+        final String path = req.uri();
+        if ("/healthcheck".equals(path)) {
+            sendHttpResponse(req, ctx, OK);
+        }
+        else if (pushConnectionPath.equals(path)) {
+            // CSRF protection
+            final String origin = req.headers().get(HttpHeaderNames.ORIGIN);
+            if ((origin == null) || (!origin.toLowerCase().endsWith(originDomain))) {
+                logger.error("Invalid Origin header {} in WebSocket upgrade request", origin);
+                sendHttpResponse(req, ctx, BAD_REQUEST);
+            }
+            else if (isDelayedAuth(req, ctx)) {
+                // client auth will happen later, continue with WebSocket upgrade handshake
+                ctx.fireChannelRead(req.retain());
+            } else {
+                final PushUserAuth authEvent = doAuth(req);
+                if (authEvent.isSuccess()) {
+                    ctx.fireChannelRead(req.retain()); // continue with WebSocket upgrade handshake
+                    ctx.fireUserEventTriggered(authEvent);
+                } else {
+                    logger.warn("Auth failed: {}", authEvent.statusCode());
+                    sendHttpResponse(req, ctx, HttpResponseStatus.valueOf(authEvent.statusCode()));
+                }
+            }
+        }
+        else {
+            sendHttpResponse(req, ctx, NOT_FOUND);
+        }
+    }
+
+    protected final Cookies parseCookies(FullHttpRequest req) {
+        final Cookies cookies = new Cookies();
+        final String cookieStr = req.headers().get(HttpHeaderNames.COOKIE);
```

Comment:
Hmm..doesn't this return a list when there are multiple Cookies involved?

Reply:
I believe SET-COOKIE and be multiple but when user agent sends cookies back it is supposed to merge them all in single COOKIE header? This is what is running in production today...

Reply:
Hmm..yeah, it's a bit unclear..most of the other code in Zuul that deals with Cookies, seems to use _getAll()_ and loop through them.

FWIW, as you said, the RFC does explicitly not allow multiple Cookies in the request..

https://tools.ietf.org/html/rfc6265#section-5.4
.."When the user agent generates an HTTP request, the user agent MUST
   NOT attach more than one Cookie header field...."

HTTP/2 allows multiple Cookie fields via HPACK (for better compression efficiency), but, does state that the HPACK decoder must concatenate all such fields into a single Header before passing it to a HTTP/1.1 context..

https://tools.ietf.org/html/rfc7540#section-8.1.2.5